File: PS_Study/BOJ/1915.py
N, M = map(int, input().split())
A = [[0 for i in range(M+1)] for j in range(N+1)]
DP = [[0 for i in range(M+1)] for j in range(N+1)]

for i in range(N):
    for idx, j in enumerate(list(map(int, list(input())))):
        A[i+1][idx+1] = j

for i in range(1, N+1):
    for j in range(1, M+1):
        if A[i][j]:
            DP[i][j] = min(DP[i-1][j], DP[i-1][j-1], DP[i][j-1]) + 1

print(max([max(i) for i in DP])**2)

File: PS_Study/1915.cpp
#include <cstdio>
#include <algorithm>
using namespace std;

int A[1005][1005], DP[1005][1005];

int main(void)
{
	int n, m, i, j, mx = 0;
	//freopen("input.txt", "rt", stdin);
	scanf("%d %d", &n, &m);
	for (i = 1; i <= n; i++) 
		for (j = 1; j <= m; j++) 
			scanf("%1d", &A[i][j]);

	for (i = 1; i <= n; i++) {
		for (j = 1; j <= m; j++) {
			if (A[i][j] == 1) {
				DP[i][j] = min(min(DP[i][j - 1], DP[i - 1][j - 1]), DP[i - 1][j]) + 1;
				mx = max(mx, DP[i][j]);
			}
		}
	}

	printf("%d", mx * mx);

	return 0;
}